File: app/src/main/java/com/tokyonth/installer/apk/ParseApkTask.java
package com.tokyonth.installer.apk;

import android.content.Context;
import android.content.pm.ActivityInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.content.pm.PermissionGroupInfo;
import android.content.pm.PermissionInfo;
import android.net.Uri;
import android.os.Handler;
import android.util.AndroidRuntimeException;

import com.tokyonth.installer.bean.ApkInfoBean;
import com.tokyonth.installer.bean.permissions.PermInfoBean;
import com.tokyonth.installer.utils.FileProviderPathUtil;
import com.tokyonth.installer.utils.ParsingContentUtil;

import java.io.File;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

public abstract class ParseApkTask extends Thread {

    private Uri uri;
    private Handler handler;
    private Context context;
    private String referrer;
    private PackageManager packageManager;
    private CommanderCallback commanderCallback;

    private PermInfoBean permInfo;
    private ApkInfoBean mApkInfo;

    void StartParseApkTask(Uri uri, Context context, Handler handler, CommanderCallback commanderCallback, String referrer) {
        this.uri = uri;
        this.handler = handler;
        this.context = context;
        this.referrer = referrer;
        this.commanderCallback = commanderCallback;
        packageManager = context.getPackageManager();
    }

    protected abstract void setApkInfo(ApkInfoBean mApkInfo);

    protected abstract void setPermInfo(PermInfoBean permInfo);

    @Override
    public void run() {
        super.run();
        try {
            handler.post(() -> commanderCallback.onStartParseApk(uri));
            mApkInfo = new ApkInfoBean();
            permInfo = new PermInfoBean();

            File QueryContent = ParsingContentUtil.getFile(context, uri);
            String apkSourcePath = (QueryContent == null) ?
                    FileProviderPathUtil.getFilePath(context, uri, referrer) : QueryContent.getPath();
            mApkInfo.setApkFile(new File(apkSourcePath));

            PackageInfo pkgInfo = packageManager.getPackageArchiveInfo(mApkInfo.getApkFile().getPath(), PackageManager.GET_ACTIVITIES);
            if (pkgInfo != null) {
                pkgInfo.applicationInfo.sourceDir = mApkInfo.getApkFile().getPath();
                pkgInfo.applicationInfo.publicSourceDir = mApkInfo.getApkFile().getPath();
                mApkInfo.setAppName(packageManager.getApplicationLabel(pkgInfo.applicationInfo).toString());
                mApkInfo.setPackageName(pkgInfo.applicationInfo.packageName);
                mApkInfo.setVersionName(pkgInfo.versionName);
                mApkInfo.setVersionCode(pkgInfo.versionCode);
                mApkInfo.setIcon(pkgInfo.applicationInfo.loadIcon(packageManager));

                ArrayList<String> activity_list = new ArrayList<>();
                if (pkgInfo.activities != null) {
                    for (ActivityInfo activity : pkgInfo.activities) {
                        activity_list.add(activity.name);
                    }
                    mApkInfo.setActivities(activity_list);
                }
                try {
                    PackageInfo installedPkgInfo = packageManager.getPackageInfo(mApkInfo.getPackageName(), PackageManager.GET_CONFIGURATIONS);
                    mApkInfo.setInstalledVersionName(installedPkgInfo.versionName);
                    mApkInfo.setInstalledVersionCode(installedPkgInfo.versionCode);
                    mApkInfo.setHasInstalledApp(true);
                } catch (PackageManager.NameNotFoundException e) {
                    e.printStackTrace();
                    mApkInfo.setHasInstalledApp(false);
                }
                PackageInfo PermPkgInfo = packageManager.getPackageArchiveInfo(mApkInfo.getApkFile().getPath(), PackageManager.GET_PERMISSIONS);
                assert PermPkgInfo != null;
                mApkInfo.setPermissions(PermPkgInfo.requestedPermissions);
                List<String> str_list = new ArrayList<>();
                if (PermPkgInfo.requestedPermissions != null) {
                    Collections.addAll(str_list, PermPkgInfo.requestedPermissions);
                    getPermissionInfo(str_list);
                }
            }
            handler.post(() -> commanderCallback.onApkParsed(mApkInfo));
            setApkInfo(mApkInfo);
        } catch (Exception e) {
            handler.post(() -> commanderCallback.onApkParsed(null));
            e.printStackTrace();
            throw new AndroidRuntimeException(e);
        }
    }

    private void getPermissionInfo(List<String> permission) {
        ArrayList<String> Group = new ArrayList<>();
        ArrayList<String> Label = new ArrayList<>();
        ArrayList<String> Description = new ArrayList<>();
        for (String str : permission) {
            try {
                PermissionInfo permissionInfo = packageManager.getPermissionInfo(str, 0);
                assert permissionInfo.group != null;
                PermissionGroupInfo permissionGroupInfo = packageManager.getPermissionGroupInfo(permissionInfo.group, 0);
                Group.add(permissionGroupInfo.loadLabel(packageManager).toString());

                String permissionLabel = permissionInfo.loadLabel(packageManager).toString();
                Label.add(permissionLabel);

                String permissionDescription = Objects.requireNonNull(permissionInfo.loadDescription(packageManager)).toString();
                Description.add(permissionDescription);

            } catch (PackageManager.NameNotFoundException e) {
                Description.add(null);
                Label.add(null);
                Group.add(null);
            }
        }
        permInfo.setPermissionDescription(Description);
        permInfo.setPermissionGroup(Group);
        permInfo.setPermissionLabel(Label);
        setPermInfo(permInfo);
    }

}

File: app/src/main/java/com/tokyonth/installer/install/ParseApkTask.kt
package com.tokyonth.installer.install

import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.system.Os
import com.tokyonth.installer.App
import com.tokyonth.installer.data.ApkInfoEntity
import com.tokyonth.installer.data.PermissionInfoEntity
import com.tokyonth.installer.utils.PackageUtils
import com.tokyonth.installer.utils.drawable2Bitmap
import com.tokyonth.installer.utils.path.UriPath
import com.tokyonth.installer.utils.path.copyFromStreamUri
import java.io.File
import java.util.zip.ZipFile

class ParseApkTask(
    private var uri: Uri,
    private var referrer: String,
    private var parse: (ApkInfoEntity?) -> Unit
) {

    private val context = App.context

    private val packageManager = context.packageManager

    fun startParseApkTask() {
        packageManager.packageInstaller
        getUriPathByFd {
            if (File(it).exists()) {
                parseApkFile(it, false)
            } else {
                UriPath.getFile(App.context, referrer,  uri).let { it1 ->
                    val name = uri.authority + ".apk"
                    if (it1 == null) {
                        val path = uri.copyFromStreamUri(name).path
                        parseApkFile(path, true)
                    } else {
                        parseApkFile(it1.path, false)
                    }
                }
            }
        }
    }

    private fun getUriPathByFd(success: (String) -> Unit) {
        try {
            App.context.contentResolver.openFileDescriptor(uri, "rw")?.use {
                val fdLink = "/proc/self/fd/${it.fd}"
                val path = Os.readlink(fdLink)
                success.invoke(path)
            }
        } catch (e: Exception) {
            success.invoke("")
        }
    }

    private fun parseApkFile(path: String?, isFake: Boolean) {
        if (path == null) {
            parse.invoke(null)
        } else {
            val apkInfo = ApkInfoEntity().apply {
                filePath = path
                isFakePath = isFake
            }
            parseApkAbi(apkInfo)
            parseApkVersion(apkInfo)
            parseApkInstalled(apkInfo)
            parseApkActivity(apkInfo)
            parseApkPermission(apkInfo)
            parse.invoke(apkInfo)
        }
    }

    private fun parseApkAbi(apkInfo: ApkInfoEntity) {
        val zip = ZipFile(apkInfo.filePath)
        val is64Bit = zip.entries().asSequence().filter {
            it.name.startsWith("lib/arm64-v8a")
        }.toMutableList().isNotEmpty()
        apkInfo.isArm64 = is64Bit
    }

    private fun parseApkVersion(apkInfo: ApkInfoEntity) {
        packageManager.getPackageArchiveInfo(apkInfo.filePath, 0)?.let {
            it.applicationInfo.sourceDir = apkInfo.filePath
            it.applicationInfo.publicSourceDir = apkInfo.filePath
            apkInfo.appName = packageManager.getApplicationLabel(it.applicationInfo).toString()
            apkInfo.packageName = it.applicationInfo.packageName
            apkInfo.versionName = it.versionName
            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.P) {
                apkInfo.versionCode = it.versionCode
            } else {
                apkInfo.versionCode = it.longVersionCode.toInt()
            }
            apkInfo.icon = it.applicationInfo.loadIcon(
                packageManager
            ).drawable2Bitmap()
        }
    }

    private fun parseApkInstalled(apkInfo: ApkInfoEntity) {
        if (PackageUtils.isAppClientAvailable(context, apkInfo.packageName)) {
            packageManager.getPackageInfo(apkInfo.packageName, 0)?.let {
                apkInfo.installedVersionName = it.versionName
                apkInfo.installedVersionCode =
                    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.P) {
                        it.versionCode
                    } else {
                        it.longVersionCode.toInt()
                    }
                apkInfo.isHasInstalledApp = true
            }
        }
    }

    private fun parseApkActivity(apkInfo: ApkInfoEntity) {
        packageManager.getPackageArchiveInfo(
            apkInfo.filePath,
            PackageManager.GET_ACTIVITIES
        )?.activities?.let { it ->
            apkInfo.activities = it.map {
                it.name
            }.toMutableList()
        }
    }

    private fun parseApkPermission(apkInfo: ApkInfoEntity) {
        packageManager.getPackageArchiveInfo(
            apkInfo.filePath,
            PackageManager.GET_PERMISSIONS
        )?.requestedPermissions?.let {
            apkInfo.permissions = getPermissionInfo(it)
        }
    }

    private fun getPermissionInfo(permission: Array<String>): MutableList<PermissionInfoEntity> {
        val list = mutableListOf<PermissionInfoEntity>()
        for (str in permission) {
            val item = PermissionInfoEntity()
            try {
                item.permissionName = str
                packageManager.getPermissionInfo(str, 0).run {
                    group.let {
                        if (it != null) {
                            item.permissionGroup = it
                        }
                    }
                    loadLabel(packageManager).let {
                        item.permissionLabel = it.toString()
                    }
                    loadDescription(packageManager).let {
                        if (it != null) {
                            item.permissionDesc = it.toString()
                        }
                    }
                }
            } catch (e: PackageManager.NameNotFoundException) {
                e.printStackTrace()
            }
            list.add(item)
        }
        return list
    }

}
